### ISSUE TITLE
Segmentation and Tracking reduce fps while streaming

### ISSUE THREAD

**Issue by AkshayLaddha943** [ultralytics/ultralytics#21548]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi all,

I have been working on sending a live stream from my Jetson orin to the GCS using ffmpeg. I am using YOLOv11n for detection which works fine with a minimal delay and an acceptable fps of 25-30. But when I include segmentation and tracking in my code, the fps drops down to 7-8 increasing the lag. 

I am using Yolo11n-seg.pt file along with BotSORT for tracking. Can anyone suggest any improvements or optimizes for improving the fps. Here is my code - 

```

import cv2
import subprocess
import time
from pymavlink import mavutil
from datetime import datetime
import torch
import numpy as np
import os
import csv
import math
from ultralytics import YOLO
from ultralytics.utils.plotting import Annotator, colors 

# === YOLOv8 Setup ===
model = YOLO('yolo11n-seg.pt')  # Segmentation model
device = 'cuda' if torch.cuda.is_available() else 'cpu'
print(f"✅ Using device: {device}")

# Define the class ID for "person" (COCO dataset class ID for person is 0)
PERSON_CLASS_ID = 0

# === MAVLink Connection ===
mav = mavutil.mavlink_connection("/dev/ttyACM0", baud=115200)
mav.wait_heartbeat()
print("✅ MAVLink: Connected to FCU")

# === Camera Setup ===
cap = cv2.VideoCapture("/dev/video0")
cap.set(cv2.CAP_PROP_FRAME_WIDTH, 640)
cap.set(cv2.CAP_PROP_FRAME_HEIGHT, 480)
cap.set(cv2.CAP_PROP_FPS, 30)

# Check if the camera opened successfully
if not cap.isOpened():
    print("Error: Could not open camera.")
    exit()

# === RTSP Streaming ===
ffmpeg_cmd = [
    "ffmpeg", "-y", "-f", "rawvideo", "-vcodec", "rawvideo", "-pix_fmt", "bgr24",
    "-s", "640x480", "-r", "20", "-i", "-",
    "-c:v", "libx264", "-preset", "ultrafast", "-tune", "zerolatency",
    "-f", "rtsp", "-rtsp_transport", "udp", "rtsp://192.168.0.130:8554/stream1"
]
ffmpeg_proc = subprocess.Popen(ffmpeg_cmd, stdin=subprocess.PIPE)

# === Logging Setup ===
os.makedirs("logged_frames", exist_ok=True)
os.makedirs("logged_telemetry", exist_ok=True)
telemetry_log_file = os.path.join("logged_telemetry", f"target_object_log_{datetime.now().strftime('%Y%m%d_%H%M%S')}.csv")
with open(telemetry_log_file, 'w', newline='') as f:
    writer = csv.writer(f)
    writer.writerow(["Timestamp", "Object", "Target_Latitude", "Target_Longitude", "Global_Latitude", "Global_Longitude", "Altitude", "Yaw", "Pixel_X", "Pixel_Y"])

# === Pixel to GPS Conversion Function ===
def pixel_to_geo(x, y, lat_c, lon_c, h):
    hfov_rad = math.radians(81)
    vfov_rad = math.radians(57)
    image_width = 640  # Match camera resolution
    image_height = 480
    IFOV_h = hfov_rad / image_width
    IFOV_v = vfov_rad / image_height

    R = 6371000  # Earth radius (meters)
    center_x = image_width / 2
    center_y = image_height / 2

    delta_theta_h = (x - center_x) * IFOV_h
    delta_theta_v = (y - center_y) * IFOV_v

    d_x = h * math.tan(delta_theta_h)
    d_y = h * math.tan(delta_theta_v)

    delta_lat = (d_y / R) * (180 / math.pi)
    delta_lon = (d_x / (R * math.cos(math.radians(lat_c)))) * (180 / math.pi)

    lat_target = lat_c + delta_lat
    lon_target = lon_c + delta_lon

    return lat_target, lon_target

# === Runtime State ===
last_alt = last_lat = last_lon = last_yaw = "N/A"
frame_count = 0
log_interval = 10

# === Main Loop ===
while True:
    # Update MAVLink telemetry
    while True:
        msg = mav.recv_match(blocking=False)
        if not msg:
            break
        if msg.get_type() == "GLOBAL_POSITION_INT":
            last_alt = f"{msg.relative_alt / 1000.0:.2f}m"
            last_lat = f"{msg.lat / 1e7:.6f}"
            last_lon = f"{msg.lon / 1e7:.6f}"
        elif msg.get_type() == "ATTITUDE":
            last_yaw = f"{msg.yaw * (180.0 / math.pi):.2f}"

    # Capture frame
    ret, frame = cap.read()
    if not ret:
        print("⚠️ Frame capture failed.")
        continue

    # Process detections and annotations
    objects_in_frame = []

     # Create an annotator to draw bounding boxes and masks
    annotator = Annotator(frame, line_width=2)

    results = model.track(frame, persist=True, tracker="botsort.yaml", device=device, classes=[PERSON_CLASS_ID], verbose=False)

    if results and results[0].masks is not None:
        person_mask = results[0].boxes.cls == PERSON_CLASS_ID
        boxes = results[0].boxes[person_mask]
        masks = results[0].masks[person_mask]
        confs = results[0].boxes.conf[person_mask]
        track_ids = results[0].boxes.id[person_mask] if results[0].boxes.id is not None else None

        im_gpu = torch.from_numpy(frame).to(device).permute(2, 0, 1).float() / 255.0

        mask_colors = [colors(int(PERSON_CLASS_ID), bgr=True) for _ in range(len(masks))]

        # Draw segmentation masks
        annotator.masks(
            masks=masks.data,
            colors=mask_colors,
            im_gpu=im_gpu,
            alpha=0.5,
            retina_masks=False
        )
        
        for i, (box, conf, mask) in enumerate(zip(boxes.xyxy, boxes.conf, masks.xy)):
            xyxy = box.cpu().numpy().astype(int)
            x_center = int((xyxy[0] + xyxy[2]) / 2)
            y_center = int((xyxy[1] + xyxy[3]) / 2)
            track_id = int(track_ids[i]) if track_ids is not None and len(track_ids) > i else -1
            try:
                alt = float(last_alt.replace('m', ''))
                lat_c = float(last_lat)
                lon_c = float(last_lon)
                lat_obj, lon_obj = pixel_to_geo(x_center, y_center, lat_c, lon_c, alt)
                label = f"Person ID {track_id} {conf:.2f} ({lat_obj:.6f}, {lon_obj:.6f})"
                objects_in_frame.append(("Person", lat_obj, lon_obj, x_center, y_center))
            except Exception as e:
                print(f"❌ Geolocation error: {e}")
                label = f"Person {conf:.2f}"

            # Get color for person class
            color = colors(int(PERSON_CLASS_ID), True)
            txt_color = annotator.get_txt_color(color)

            # Annotate with segmentation mask and bounding box
            annotator.box_label(
                box=xyxy,
                label=label,
                color=color,
                txt_color=txt_color
            )

    # Get annotated frame
    annotated_frame = annotator.result()

    # Overlay telemetry
    timestamp = datetime.now().strftime("%H:%M:%S")
    overlay = f"[{timestamp}] Alt: {last_alt} | Yaw: {last_yaw} | Lat: {last_lat} | Lon: {last_lon}"
    cv2.putText(annotated_frame, overlay, (10, 50), cv2.FONT_HERSHEY_SIMPLEX, 0.4, (255, 255, 255), 2)

    # Stream frame
    ffmpeg_proc.stdin.write(annotated_frame.tobytes())

    # Logging
    frame_count += 1
    if frame_count % log_interval == 0:
        log_time = datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
        cv2.imwrite(f"logged_frames/frame_{log_time}.jpg", annotated_frame)
        with open(telemetry_log_file, 'a', newline='') as f:
            writer = csv.writer(f)
            for obj_name, obj_lat, obj_lon, x_center, y_center in objects_in_frame:
                writer.writerow([log_time, obj_name, obj_lat, obj_lon, last_lat, last_lon, last_alt, last_yaw, x_center, y_center])

# Release resources
cap.release()
ffmpeg_proc.stdin.close()
ffmpeg_proc.terminate()
```

I tried converting the .pt file to .engine,but it starts giving me resolution issue.Any help would be appreciated

**Comment by UltralyticsAssistant** [comment]
👋 Hello @AkshayLaddha943, thank you for sharing the details of your workflow and for your interest in Ultralytics 🚀! We recommend visiting the [Docs](https://docs.ultralytics.com/) for helpful [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples, as many common questions are already addressed there.

Since this is a performance-related 🐛 Bug Report/Question, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) if you haven't already. This will help us and the community to better understand and reproduce your issue.

If you have any additional information—such as a short test script, logs, or specific error messages from your .engine conversion—please include those as well.

Join our community for deeper discussions and real-time help:
- [Discord](https://discord.com/invite/ultralytics) 🎧 for chat
- [Discourse](https://community.ultralytics.com/) for detailed Q&A
- [Subreddit](https://www.reddit.com/r/Ultralytics/) for community threads

## Upgrade

To ensure your environment is up to date, please upgrade to the latest `ultralytics` package (along with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml)) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of these up-to-date, verified environments (with all dependencies like [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

This is an automated response to help you get started. An Ultralytics engineer will follow up with you soon for more in-depth assistance. Thank you for your patience and for contributing to the Ultralytics community! 😊

**Comment by Y-T-G** [comment]
If you want faster FPS, you should use TensorRT, particularly exporting with `half=True`. What do you mean by "resolution issue"?

And you can use `bytetrack.yaml` instead of `botsort.yaml` as `tracker` because it's faster.

You're already filtering person class by passing `classes=[0]` to `model.predict()`. There's no need to perform masking again with class ID.

And you shouldn't be writing to CSV in the same thread. That would slow down the inference. It's an I/O operation. If you want to write to CSV, you should do it in a separate thread.

You're doing too many operations in the inference thread that will slow down inference.

**Comment by Chaitali2209** [comment]
> If you want faster FPS, you should use TensorRT, particularly exporting with `half=True`. What do you mean by "resolution issue"?
> 
> And you can use `bytetrack.yaml` instead of `botsort.yaml` as `tracker` because it's faster.
> 
> You're already filtering person class by passing `classes=[0]` to `model.predict()`. There's no need to perform masking again with class ID.
> 
> And you shouldn't be writing to CSV in the same thread. That would slow down the inference. It's an I/O operation. If you want to write to CSV, you should do it in a separate thread.
> 
> You're doing too many operations in the inference thread that will slow down inference.



Thanks a lot, exporting model with half=True helped and creating  separate thread for csv logging improved the fps to 25.

**Comment by glenn-jocher** [comment]
Great to hear the optimizations worked! For additional performance gains on Jetson, you might also consider setting `with_reid=False` in your tracker configuration since ReID adds computational overhead and may not be necessary for your use case. The [tracking documentation](https://docs.ultralytics.com/modes/track/) shows how ReID can be disabled to minimize performance impact while still maintaining effective tracking.
